system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1276,416 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+                                                   g_strdup (g_dbus_connection_get_unique_name (connection)),
+                                                   g_free);
+
+  if (socketpair (AF_UNIX, SOCK_SEQPACKET, 0, sockets) == -1)
+    {
+      glnx_throw_errno_prefix (error, "Failed to get a socketpair");
+      return NULL;
+    }
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_set_child_setup (launcher, revokefs_fuse_backend_child_setup, passwd, NULL);
+  g_subprocess_launcher_take_fd (launcher, sockets[0], 3);
+  fcntl (sockets[1], F_SETFD, FD_CLOEXEC);
+  pull->client_socket = sockets[1];
+
+  pull->revokefs_backend = g_subprocess_launcher_spawn (launcher,
+                                                        error,
+                                                        "revokefs-fuse",
+                                                        "--backend",
+                                                        "--socket=3", src, NULL);
+  if (pull->revokefs_backend == NULL)
+    return NULL;
+
+  return g_steal_pointer (&pull);
+}
+
+static gboolean
+check_for_repo_containing_partial_pull (GFile          *cache_dir,
+                                        const gchar    *commit,
+                                        gchar         **out_repo_location,
+                                        struct passwd  *passwd)
+{
+  g_autoptr(GFileEnumerator) enumerator = NULL;
+  GFileInfo *file_info = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *cache_dirname = g_file_get_path (cache_dir);
+  struct stat st_buf;
+  const gchar *name;
+
+  if (stat (cache_dirname, &st_buf) == 0 &&
+      /* should be owned by SYSTEM_HELPER_USER */
+      st_buf.st_uid == passwd->pw_uid &&
+      /* should not be world-writeable */
+      (st_buf.st_mode & 0022) == 0)
+    g_debug ("Cache directory %s status valid", cache_dirname);
+  else
+    {
+      g_warning ("Cache directory %s status invalid", cache_dirname);
+      return FALSE;
+    }
+
+  enumerator = g_file_enumerate_children (cache_dir,
+                                          G_FILE_ATTRIBUTE_STANDARD_NAME,
+                                          G_FILE_QUERY_INFO_NONE, NULL, &error);
+  if (enumerator == NULL)
+    {
+      g_warning ("Failed to enumerate %s: %s", cache_dirname, error->message);
+      return FALSE;
+    }
+
+   while (TRUE)
+    {
+      if (!g_file_enumerator_iterate (enumerator, &file_info, NULL, NULL, &error))
+        {
+          g_warning ("Error while iterating %s: %s", cache_dirname, error->message);
+          break;
+        }
+
+      if (file_info == NULL)
+        break;
+
+      name = g_file_info_get_name (file_info);
+      if (g_file_info_get_file_type (file_info) == G_FILE_TYPE_DIRECTORY &&
+          g_str_has_prefix (name, "repo-"))
+        {
+          g_autofree gchar *commit_partial_basename = g_strdup_printf ("%s.commitpartial", commit);
```

Comment:
Need to handle commit not being specified here.

Reply:
Well, yeah commit can be NULL as per your previous comment. 
In that case, we cannot re-use this directory right? or Should we ? :thinking: 
What would be the key identifier that the partial repo belongs to this pull ...

Reply:
The way ostree normally reuse pull directories is it just picks any existing one. It may be the case that it is the wrong one, but normally when you get a failure you immediately retry. This way old failed directories are automatically cleaned up too. See here for this code: 
https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-repo.c#L5883

I'm not sure this complexity for finding the "right" directory to reuse is worth it in practice.

Reply:
Stripped per-ref repo and pull recovery(for that $ref) approach in 071a00f57a9ac as discussed on IRC. We now pick-up any cache-directory available in `repo/tmp`.